File: java/src/test/java/sammancoaching/SalaryPensionContributionCalculationTest.java
package sammancoaching;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.HashMap;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class SalaryPensionContributionCalculationTest {

    public static final double NORMAL_BASE_SALARY = 60000.0;
    private final SalaryContributionPercentages fakePercentages = FakePercentages.getStandardValues();

    @Test
    public void belowZeroSalary_fails() {
        assertThrows(IllegalArgumentException.class, ()-> PensionContributionCalculator.calculatePensionContribution(
                BigDecimal.valueOf(-1), -1, null, fakePercentages));
    }

    @Test
    public void juniorEmployeeWithNoTenure_BasicContribution() {
        BigDecimal annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY);
        int tenure = 0;
        SeniorityLevel seniority = new JuniorEmployee();

        var actualContribution = PensionContributionCalculator.calculatePensionContribution(
                annualSalary, tenure, seniority, fakePercentages);

        assertEquals(3000.0,
                actualContribution.doubleValue(), 0.001);
    }

    @Test
    public void midLevelRecentHire_MediumContribution() {
        BigDecimal annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY);
        int tenure = 0;
        SeniorityLevel seniority = new MidLevel();

        var actualContribution = PensionContributionCalculator.calculatePensionContribution(
                annualSalary, tenure, seniority, fakePercentages);

        assertEquals(4800.0,
                actualContribution.doubleValue(), 0.001);
    }

    @Test
    public void midLevelMediumTenure_MediumLargeContribution() {
        BigDecimal annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY);
        int tenure = 5;
        SeniorityLevel seniority = new MidLevel();

        var actualContribution = PensionContributionCalculator.calculatePensionContribution(
                annualSalary, tenure, seniority, fakePercentages);

        assertEquals(6000.0,
                actualContribution.doubleValue(), 0.001);
    }

    @Test
    public void leadershipWithLongTenure_MaximumContribution() {
        BigDecimal annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY);
        int tenure = 25;
        SeniorityLevel seniority = new LeadershipTeam();

        var actualContribution = PensionContributionCalculator.calculatePensionContribution(
                annualSalary, tenure, seniority, fakePercentages);

        assertEquals(6600.0,
                actualContribution.doubleValue(), 0.001);
    }

}

File: kotlin/src/test/kotlin/sammancoaching/SalaryPensionContributionCalculationTest.kt
package sammancoaching

import java.math.BigDecimal
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class SalaryPensionContributionCalculationTest {

    val NORMAL_BASE_SALARY: Double = 60000.0
    private val fakePercentages: SalaryContributionPercentages? = FakePercentages.getStandardValues()

    @Test
    fun belowZeroSalary_fails() {
        assertFailsWith(IllegalArgumentException::class, {
            PensionContributionCalculator.calculatePensionContribution(
                BigDecimal.valueOf(-1), -1, JuniorEmployee(), fakePercentages!!
            )
        })
    }

    @Test
    fun juniorEmployeeWithNoTenure_BasicContribution() {
        val annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY)
        val tenure = 0
        val seniority: SeniorityLevel = JuniorEmployee()

        val actualContribution = PensionContributionCalculator.calculatePensionContribution(
            annualSalary, tenure, seniority, fakePercentages!!
        )

        assertEquals(
            3000.0,
            actualContribution.toDouble(), 0.001
        )
    }

    @Test
    fun midLevelRecentHire_MediumContribution() {
        val annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY)
        val tenure = 0
        val seniority: SeniorityLevel = MidLevel()

        val actualContribution = PensionContributionCalculator.calculatePensionContribution(
            annualSalary, tenure, seniority, fakePercentages!!
        )

        assertEquals(
            4800.0,
            actualContribution.toDouble(), 0.001
        )
    }

    @Test
    fun midLevelMediumTenure_MediumLargeContribution() {
        val annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY)
        val tenure = 5
        val seniority: SeniorityLevel = MidLevel()

        val actualContribution = PensionContributionCalculator.calculatePensionContribution(
            annualSalary, tenure, seniority, fakePercentages!!
        )

        assertEquals(
            6000.0,
            actualContribution.toDouble(), 0.001
        )
    }

    @Test
    fun leadershipWithLongTenure_MaximumContribution() {
        val annualSalary = BigDecimal.valueOf(NORMAL_BASE_SALARY)
        val tenure = 25
        val seniority: SeniorityLevel = LeadershipTeam()

        val actualContribution = PensionContributionCalculator.calculatePensionContribution(
            annualSalary, tenure, seniority, fakePercentages!!
        )

        assertEquals(
            6600.0,
            actualContribution.toDouble(), 0.001
        )
    }
}